### DISCUSSION TITLE
Row expansion question

### DISCUSSION THREAD

**Discussion by akramer24** [perspective-dev/perspective#1774] [Q&A] [answered]
Hello,

I am experimenting with `perspective-python` as a backend solution for a pivot table. I'm having some trouble understanding how row expansion works. When I pivot a view and attempt to expand a row, `get_row_expanded` subsequently tells me the row is not expanded and I also do not see the expanded row's children in `view.to_records()`. A simple repro is below. Any guidance you could provide would be greatly appreciated.

```python
my_table = perspective.Table(
    pd.DataFrame(
        [
            ["Fruit", "Apple", 10],
            ["Fruit", "Orange", 5],
            ["Fruit", "Strawberry", 12],
            ["Vegetable", "Potato", 8],
            ["Vegetable", "Carrot", 4],
        ],
        columns=["Category", "Item", "Quantity"],
    ),
    index="index",
)
pivoted = my_table.view(group_by=["Category"], columns=["Quantity"])
print(
    pivoted.to_records()
)  # [{'__ROW_PATH__': [], 'Quantity': 39}, {'__ROW_PATH__': ['Fruit'], 'Quantity': 27}, {'__ROW_PATH__': ['Vegetable'], 'Quantity': 12}]
print(pivoted.get_row_expanded(1))  # False
print(pivoted.get_row_expanded(2))  # False
pivoted.expand(1)
print(
    pivoted.to_records()
)  # [{'__ROW_PATH__': [], 'Quantity': 39}, {'__ROW_PATH__': ['Fruit'], 'Quantity': 27}, {'__ROW_PATH__': ['Vegetable'], 'Quantity': 12}]
print(pivoted.get_row_expanded(1))  # False
```

**Comment by texodus** [comment, marked as answer]
Your problem isn't with row expansion, rather this is not how group-by works.  You can't expand the `"Fruit"` row because this is a leaf node, there are no children because you've only applied a single pivot level.  Please [check the documentation for `group_by`](https://perspective.finos.org/docs/md/view.html#group-by), or any of the interactive examples on [the perspective docs site](https://perspective.finos.org/) that feature group-by operations.

If you want a pivot table, use `perspective-viewer` - there is nothing in the `perspective-python` API which is not 1:1 already implemented in the `perspective-viewer` UI.  If you want a master-detail view, [here is an example](https://bl.ocks.org/texodus/54857753e5642799240ff1ecd161175a).

**Reply by akramer24** [reply]
Thank you for the quick response. This was helpful. I didn't realize I was using `group_by` incorrectly.
